numcodecs/bitround.py

```diff
@@ -0,0 +1,31 @@
+import numpy as np
+
+
+from .abc import Codec
+from .compat import ensure_ndarray, ndarray_copy
+
+
+class BitRound(Codec):
+    codec_id = 'bitround'
+
+    def __init__(self, keepbits: int):
+        self.keepbits = keepbits
+
+    def encode(self, buf):
+        # TODO: figure out if we need to make a copy
+        # Currently this appears to be overwriting the input buffer
+        # Is that the right behavior?
+        a = ensure_ndarray(buf).view()
+        assert a.dtype == np.float32
+        b = a.view(dtype=np.int32)
+        maskbits = 23 - self.keepbits
+        mask = (0xFFFFFFFF >> maskbits) << maskbits
+        half_quantum1 = (1 << (maskbits - 1)) - 1
```

Comment:
This section will error with `ValueError: negative shift count` if `keepbits=23`. That causes the `test_no_rounding` test to error.

Reply:
This is a feature. half_quantum1 is indeed inexpressible if your quantum is the least significant bit... I would just add return from the beginning or not call the sub at all if keepbits == 23.  If  keepbits > 23 (or  < 0) you might wish to raise an exception... 
I wonder if the test is needed at all...

Reply:
I find a condition like `keepbits >= 0` reasonable, but note that `keepbits < 0` within reasonable limits (see below) shouldn't cause problems as this will just round the exponent bits. E.g.
```julia
julia> using BitInformation
julia> round(4f0,-1)
2.0f0
```
So in this format only ...,0.125,0.5,2,8,... are representable. However, as the exponent bits describe logarithmically distributed floats, round to nearest is now round to nearest in log-space. Meaning that while 4 is round down to 2 in the example above,
```julia
julia> round(4.1f0,-1)
8.0f0
```
although in lin-space 4.1 is closer to 2 than to 8. Sure, this has to be treated with caution, as NaN/Inf are defined through their exponents, meaning you end up with situations like
```julia
julia> round(NaN32,-1)
-0.0f0
julia> round(Inf32,-1)
-0.0f0
```
And the carry bit can propagate into the sign bit (for |x|>2), which is also weird
```julia
julia> round(4f0,-8)
-0.0f0
julia> round(-4f0,-8)
0.0f0
```